### PR TITLE
feat: add manacher's algorithm 

### PR DESCRIPTION
**Pull request by riti2409**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: This is an implementation of the [Manacher's Algorithm](https://en.wikipedia.org/wiki/Longest_palindromic_substring) which is used to find the longest substring in linear time.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1577"></a>

### CODE REVIEW

**Review by b30wulffz — APPROVED** [review]
Looks good to me.

File: strings/manacher_algorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    assert(strings::manacher::manacher("") == "");
+    assert(strings::manacher::manacher("abababc") == "ababa");
+    assert(strings::manacher::manacher("cbaabd") == "baab");
+    assert(strings::manacher::manacher("xacdedcax") == "xacdedcax");
+    assert(strings::manacher::manacher("xacddcax") == "xacddcax");
```
Comment (vid-7695) [inline strings/manacher_algorithm.cpp:152]: Add some test cases with Upper , Mixed, camel Case. Test Cases format seem to be repeativie.

File: strings/manacher_algorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    assert(strings::manacher::manacher("a") == "a");
+    assert(strings::manacher::manacher("ab") == "a");
+    assert(strings::manacher::manacher("abced") == "a");
```
Comment (vid-7695) [inline strings/manacher_algorithm.cpp:155]: Add a null test case, to see if that is being handled.

**Review by vid-7695 — DISMISSED** [review]

File: strings/manacher_algorithm.cpp (hunk 2 above)
Comment (riti2409) [reply strings/manacher_algorithm.cpp:155]: Handled it on line 148

**Review by shikhar-psyduck — DISMISSED** [review]
LGTM

File: strings/manacher_algorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
+/**
+ * @file
+ * @brief Implementation of the
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:3]:
```suggestion
 * @brief Implementation of
```

File: strings/manacher_algorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ * @details
+ * Manacher's Algorithm is used to find longest palindromic substring within a
+ * string in O(n) time. It exploits the property of a palindrome that its first
+ * half is symmetric to the last half, and thus if the first half is a
+ * palindrome, then last half is also a palindrome.
+ * @author [Riti Kumari](https://github.com/riti2409)
+ */
+
+#include <cassert>   // for asserting test cases
+#include <iostream>  // for IO operations
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:15]:
```suggestion
#include <cassert>   /// for assert
#include <iostream>  /// for IO operations
```

File: strings/manacher_algorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+#ifdef _MSC_VER
+#include <string>  // required for MS Visual C++
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:17]:
```suggestion
#include <string>  /// required for MS Visual C++
```

File: strings/manacher_algorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+#else
+#include <cstring>
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:19]: Add a one-line description here as in line 17.

File: strings/manacher_algorithm.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ * @brief Functions for the
+ * [Manacher's
+ * Algorithm](https://en.wikipedia.org/wiki/Longest_palindromic_substring)
+ * implementation
+ */
+namespace manacher {
+/**
+ * @brief A function that implements Manacher's algorithm
+ * @param prototype is the string where algorithm finds a palindromic substring.
+ * This string can contain any character except @ # &
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:38]:
```suggestion
 * This string can contain any character except `@` `#` `&`
```

File: strings/manacher_algorithm.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run self-test implementations
+    return 0;
+}
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:165]:
```suggestion
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work, @riti2409! Thanks for reviewing, @b30wulffz, @vid-7695, and @shikhar-psyduck. 🙂
Also, please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1577/checks?check_run_id=3464063442) warnings.

**Comment by riti2409** [comment]
@Panquesito7 I have incorporated all the suggestions. Please review.

File: strings/manacher_algorithm.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+ * Manacher's Algorithm is used to find longest palindromic substring within a
+ * string in O(n) time. It exploits the property of a palindrome that its first
+ * half is symmetric to the last half, and thus if the first half is a
+ * palindrome, then last half is also a palindrome.
+ * @author [Riti Kumari](https://github.com/riti2409)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for vector STL
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:15]:
```suggestion
#include <vector>    /// for std::vector STL
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Comment by riti2409** [comment]
@Panquesito7  Done 😀

**Review by b30wulffz — APPROVED** [review]

File: strings/manacher_algorithm.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+        std::vector<int> palindrome_max_half_length(
+            stuffed_string.size(),
+            0);  // int palindrome_max_half_length[stuffed_string.size()] = {0};
+                 // this array will consist of largest possible half length of
+                 // palindrome centered at index (say i with respect to the
+                 // stuffed string). This value will be lower bound of half
+                 // length since single character is a palindrome in itself.
+
+        int bigger_center =
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:59]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

File: strings/manacher_algorithm.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    assert(strings::manacher::manacher("cbaabd") == "baab");
+    assert(strings::manacher::manacher("DedzefDeD") == "DeD");
+    assert(strings::manacher::manacher("XZYYXXYZXX") == "YXXY");
+    assert(strings::manacher::manacher("1sm222m10abc") == "m222m");
+    assert(strings::manacher::manacher("798989591") == "98989");
+    assert(strings::manacher::manacher("xacdedcax") == "xacdedcax");
+    assert(strings::manacher::manacher("xaccax") == "xaccax");
+    assert(strings::manacher::manacher("a") == "a");
+    assert(strings::manacher::manacher("xy") == "x");
+    assert(strings::manacher::manacher("abced") == "a");
```
Comment (Panquesito7) [inline strings/manacher_algorithm.cpp:161]:
```suggestion
    assert(strings::manacher::manacher("abced") == "a");
    
    std::cout << "All tests have passed!" << std::endl;
```

Comment (riti2409) [reply strings/manacher_algorithm.cpp:162]: Done

File: strings/manacher_algorithm.cpp (hunk 10 above)
Comment (riti2409) [reply strings/manacher_algorithm.cpp]: @Panquesito7 Sure, Updated

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @riti2409! Thanks for your contribution! 😄👍🎉

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Oh, you just missed one thing: please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1577/checks?check_run_id=3466067139) warnings.

**Comment by riti2409** [comment]
Compile checks for "Awesome CI Workflow / Compile checks (windows-latest) (pull_request)" failed to run because of some server issue in the VM. Is it possible to rerun the checks?

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @riti2409! It's been great for your first contribution here! Thank you; we hope you keep contributing. 😄 👍 🎉

**Review by ayaankhan98 — APPROVED** [review]
LGTM :+1:

**Comment by riti2409** [comment]
@Panquesito7 Thanks for encouraging :)

**Pull request merged**